core/vm/runtime/circulating.cpp

```diff
@@ -0,0 +1,121 @@
+#include "vm/actor/builtin/v0/storage_power/storage_power_actor_state.hpp"
+#include "vm/actor/builtin/v2/codes.hpp"
+#include "vm/actor/builtin/v2/market/actor.hpp"
+#include "vm/actor/builtin/v2/power/actor.hpp"
+#include "vm/actor/builtin/v2/reward/actor.hpp"
+#include "vm/state/impl/state_tree_impl.hpp"
+#include "vm/version.hpp"
+
+namespace fc::vm {
+  outcome::result<TokenAmount> getLocked(StateTreePtr state_tree) {
+    auto ipld{state_tree->getStore()};
+    TokenAmount locked;
+    OUTCOME_TRY(market, state_tree->get(actor::kStorageMarketAddress));
+    if (market.code == actor::builtin::v0::kStorageMarketCodeCid
+        || market.code == actor::builtin::v2::kStorageMarketCodeCid) {
+      static_assert(std::is_same_v<actor::builtin::v0::market::State,
+                                   actor::builtin::v2::market::State>);
+      OUTCOME_TRY(
+          state, ipld->getCbor<actor::builtin::v0::market::State>(market.head));
+      locked += state.total_client_locked_collateral
+                + state.total_provider_locked_collateral
+                + state.total_client_storage_fee;
+    } else {
+      return std::errc::owner_dead;
+    }
+    OUTCOME_TRY(power, state_tree->get(actor::kStoragePowerAddress));
+    if (power.code == actor::builtin::v0::kStoragePowerCodeCid) {
+      OUTCOME_TRY(
+          state,
+          ipld->getCbor<actor::builtin::v0::storage_power::State>(power.head));
+      locked += state.total_pledge;
+    } else if (power.code == actor::builtin::v2::kStoragePowerCodeCid) {
+      OUTCOME_TRY(state,
+                  ipld->getCbor<actor::builtin::v2::power::State>(power.head));
+      locked += state.total_pledge;
+    } else {
+      return std::errc::owner_dead;
+    }
+    return locked;
+  }
+
+  outcome::result<std::shared_ptr<Circulating>> Circulating::make(
+      IpldPtr ipld, const CID &genesis) {
+    auto circulating{std::make_shared<Circulating>()};
+    OUTCOME_TRY(block, ipld->getCbor<primitives::block::BlockHeader>(genesis));
+    OUTCOME_TRYA(circulating->genesis,
+                 getLocked(std::make_shared<state::StateTreeImpl>(
+                     ipld, block.parent_state_root)));
+    return circulating;
+  }
+
+  outcome::result<TokenAmount> Circulating::circulating(
+      StateTreePtr state_tree, ChainEpoch epoch) const {
+    auto ipld{state_tree->getStore()};
+    TokenAmount vested, mined;
+
+    auto vest{[&](auto days, TokenAmount amount) {
+      ChainEpoch duration(days * vm::actor::builtin::v0::miner::kEpochsInDay);
+      auto elapsed{epoch};
+      if (epoch > version::kUpgradeIgnitionHeight) {
+        amount *= kFilecoinPrecision;
+        elapsed -= version::kUpgradeLiftoffHeight;
+      }
+      if (elapsed >= duration) {
+        vested += amount;
+      } else if (elapsed >= 0) {
+        vested += amount - (duration - elapsed) * bigdiv(amount, duration);
+      }
+    }};
+    constexpr auto sixMonths{183}, year{365};
```

Comment:
What about leap year?

Reply:
Lotus doesn't care

Reply:
Sounds not good in general, but ok for this pr